WIP: Match timestep for well info

lars-petter-hauge opened the pull request:
NB: I added the PR as it was in opm-output. Both changes in test and commit msg may stay as is.

This PR handles when opm-output takes into account changes in group policy or efficiency factor during a run. Opm-output would previously change those configurations according to when they are parsed, however as a reporting tool, the configuration should instead be retrieved from the previous time step.

Decrement timestep where necessary to avoid premature well configuration.

Consider the following:
```
.
.
.

WELLSPECS
 'W1'  'G1' ... /

TSTEP
 2*1 /

WELLSPECS
 'W1'  'G2' .../

TSTEP
 2*1 /
```
In that case the parent group of `W1` is changed after two timesteps. At timestep two the Group parent will indicate `G2`, while the `GOPT:G2` at timestep two should be reported for what has progressed up until timestep two.

totto82 commented on src/opm/output/eclipse/Summary.cpp:
```diff
@@ -775,6 +775,7 @@ inline std::vector< const Well* > find_wells( const Schedule& schedule,
     if( type == ECL_SMSPEC_GROUP_VAR ) {
         if( !schedule.hasGroup( name ) ) return {};
+        if (timestep!=0) timestep--;
 
```
What about adding a variable for this. i.e.  previous_timestep_index = timestep - 1 to avoid misuse later.

totto82 commented:
I noticed the same bug while testing a model with non constant WEFAC. I can confirm that this fixes my issue.

lars-petter-hauge commented:
I've changed as @totto82  suggested, and also added a comment. As it is now I do believe it should be fairly well documented and clear.

I thought about adding the fix to parser instead, as indicated by @joakim-hove in https://github.com/OPM/opm-output/pull/257, but adding another function for such a small and specific usage..? Or have I misunderstood how this could be handled in the parser part?

joakim-hove commented on src/opm/output/eclipse/Summary.cpp:
```diff
@@ -218,7 +218,7 @@ inline quantity rate( const fn_args& args ) {
 inline quantity flowing( const fn_args& args ) {
     const auto& wells = args.wells;
-    const auto ts = args.timestep;
+    const auto ts = args.timestep + 1;
```
Are you certain this should be `+1` - I would actually think that we want here is the raw `timestep` value; although I agree that will be a change.

joakim-hove commented on src/opm/output/eclipse/Summary.cpp:
```diff
@@ -1041,14 +1029,14 @@ void Summary::add_timestep( int report_step,
     /* Subtract one index to get well information from
      * the previous to the current timestep */
-    size_t prev_timestep = timestep > 0 ? timestep - 1 : 0;
+    const size_t timestep = report_step > 0 ? report_step - 1 : 0;
```
Please - I think this should be `sim_timestep` instead of just `timestep`.

joakim-hove commented on src/opm/output/eclipse/Summary.cpp:
```diff
@@ -1041,14 +1029,14 @@ void Summary::add_timestep( int report_step,
 
-        const auto schedule_wells = find_wells( schedule, f.first, prev_timestep, this->regionCache );
-        auto eff_factors = well_efficiency_factors( f.first, schedule, schedule_wells, prev_timestep );
+        const auto schedule_wells = find_wells( schedule, f.first, timestep, this->regionCache );
```
`timestep` -> `sim_timestep`

joakim-hove commented:
This became a bit awkward. As I see it it should be possible to be quite consistent as such:

We have two concepts: 
1. `report_step` that is the number of the file we are about to write - i.e. for instance `CASE.S$report_step` for the data in a non-unified summary file.
2. `simulation_step` or `sim_step` that is the timestep which has been effective in the simulator, when actually simulating the current results, and the value which should be used when e.g. consulting the `Schedule`object.

The input argument to the `Summary::add_timestep()`routine is `report_step`, so that routine should immediately start with:
```C++
size_t simulation_step = report_step - 1;
```
the value used in `fn_args` should be renamed `simulation_step` - and of course the simulaton_step` value should be used. As for the `prev_step` which is used for total values, that is used to look up the previous value in the generated `ecl_sum` datastructure - i.e. that is in the `report_step` value space, and I guess a rename to `prev_report_step` would clarify that.

I think test changes - i.e. using `wells.empty()` instead of explicitly checking timestep values is very good,.

lars-petter-hauge commented:
Perhaps I dived too far there.. Thanks for the clarification on time steps though - I took the liberty to add the comment in the code.

I might need to add an explanation as to why report_step is used instead of sim_step for <flowing> (which is the only one that uses report_step.

I believe the code should now be fairly well structured. The only change in the end now is that I've used int instead of size_t. Is there any reason size_t should be used? Using size_t does not work well with e.g. std::max, a function that describes neatly what we're looking for.

joakim-hove commented:
jenkins build this please

joakim-hove commented:
The final merge will be here: https://github.com/OPM/opm-common/pull/354
